### DISCUSSION TITLE
Race condition in subscription updates

### DISCUSSION THREAD

**Discussion by zenios** [urql-graphql/urql#2770] [Q&A] [answered]
### Describe the bug

Hi.

Lets say i have the following schema
```
type Query {
    user:User!
}

type Subscription {
    userUpdates:UpdatEvent!
}

type User {
    id:ID!
    balance:Float!
    notifications:Int!
}

type BalanceUpdate {
    balance:Float!
}

type NotificationCreated {
    id:ID!
}

type NotificationRemoved {
    id:ID!
}

union UpdateEvent = BalanceUpdate | NotificationCreated | NotificationRemoved
```

And the following graphcache updater
```
function userUpdates(
  data: UserUpdatesSubscription,
  args: UserUpdatesSubscriptionVariables,
  cache: Cache
) {
  const event = data.userUpdates;
            cache.updateQuery<UserQuery, UserQueryVariables>(
    {
      query: USER_QUERY,
    },
    (cachedData) => {
      if (!cachedData?.user) return cachedData;
      if (event.__typename === "BalanceUpdate")
        cachedData.user.balance = event.balance;
      else if (event.__typename === "NotificationCreated")
        cachedData.user.notifications = cachedData.user.notifications + 1;
      else if (event.__typename === "NotificationRemoved")
        cachedData.user.notifications = cachedData.user.notifications - 1;
      return cachedData;
    }
  );
}

```

If two events arrive at the same time for example BalanceUpdate and NotificationCreated some times the cached data of the updater returns previous data. One example is the following

```
user = {
  id: "test",
  balance: 10,
  notifications: 1,
}

NotificationCreated arrived

user = {
  id: "test",
  balance: 10,
  notifications: 2,
};

Balance update of value 20 arrived right after NotificationCreated

user = {
  id: "test",
  balance: 20,
  notifications: 1,
};

```

### Reproduction

no-reproduction

### Urql version

"@urql/exchange-auth": "^0.1.7",
"@urql/exchange-graphcache": "^5.0.3",
"@urql/exchange-multipart-fetch": "^0.1.14",
"urql": "^3.0.3",

### Validations

- [X] I can confirm that this is a bug report, and not a feature request, RFC, question, or discussion, for which [GitHub Discussions](https://github.com/FormidableLabs/urql/discussions) should be used
- [X] Read the [docs](https://formidable.com/open-source/urql/docs/).
- [X] Follow our [Code of Conduct](https://github.com/FormidableLabs/urql/blob/main/CODE_OF_CONDUCT.md)

**Comment by JoviDeCroock** [comment]
From a consistency point of view this makes sense, the latest mutation will return the latest data, if your server is returning stale data for the same entity I would argue there might be a consistency issue there.

When two mutations are fired we construct [layers](https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/#deterministic-cache-updates) so they are applied in the correct order, i.e. the first mutation and then the second mutation regardless of the order in which they return data. You could work around this by only asking for the data you know is changing in the mutation.

**Comment by zenios** [comment]
Hi @JoviDeCroock i think you misunderstood the issue. Never the less here is a link to reproduction

https://github.com/zenios/urql-subscriptions-issue

Sorry for not providing codesandbox. At the time i was writing the reproduction there containers where not working.

**Comment by JoviDeCroock** [comment]
I mean, from that example I can only derive that this is a bit against what GraphQL is supposed to do, you get your server-state back from your mutation and then the subscription trigger makes you increment/decrement it. Shouldn't the subscription tell you the server-state as well?

Converting this to a discussion as it's a schema design inconsistency, you are reporting the event of an increase/decrease which makes you do a manual update, which isn't something we could ever account for. You update the local-state with server-state by means of mutation resposne and then do a manual one

**Comment by zenios** [comment]
I don't think that's the case @JoviDeCroock 

Its almost identical to the thousand of examples online in where you update a messages list with new messages coming from subscriptions.

Convert notifications field to an array of notifications and subscription to send the new notifications and you will end up in the same situation.

In that case if N notifications arrive in parallel there is a chance some of them to get lost

**Comment by kitten** [comment, marked as answer]
Hiya 👋

We found a cause for this which wasn't quite obvious, let me explain.

Typically, we often see events and updaters reflecting the "true state" when it comes to GraphQL subscriptions. For instance, in the case of changing numbers, it's common to write a subscription event that'd notify the client of "You have X number of items" rather than "You have a new item".

The reason for that is that it's easy to miss events and that would then mean that queries (the GraphQL API state, basically) and the subscription state (i.e. the local state) go out of sync.

When however the subscription is written as you did, then a very particular and subtle bug occurred. When a subscription is used that continuously builds on past data, then a small bug caused it to lose data on every second write, because it was trying to prevent optimistic data from being treated as the "true data".

However, obviously, subscriptions always reflect true data, even when they're continuously accumulating new data. We filed a PR for this and this should fix this edge case. It was basically caused by a single condition being incorrectly rewritten when we added support for deferred queries (which are basically one of multiple ways to turn queries into subscriptions in GraphQL from a caching perspective)

**Comment by zenios** [comment]
Hi

Thanks for finding the issue. As said the number increase was just an
example that made it more obvious.

The same could have happened in cases we are sending new messages that
should be added to an existing list of messages.
